File: 0994_rotting_oranges/rotting_oranges.py
from collections import deque
from typing import List


def oranges_rotting(grid: List[List[int]]) -> int:
    """
    Time: O(m * n)
    Space: O(m * n)
    """
    n, m = len(grid), len(grid[0])
    rotten = deque()
    fresh = 0
    for i in range(n):
        for j in range(m):
            if grid[i][j] == 2:
                rotten.append((i, j))
            elif grid[i][j] == 1:
                fresh += 1
    minutes = 0
    while rotten:
        for _ in range(len(rotten)):
            i, j = rotten.popleft()
            for x, y in [(i + 1, j), (i - 1, j), (i, j + 1), (i, j - 1)]:
                if 0 <= x < n and 0 <= y < m and grid[x][y] == 1:
                    fresh -= 1
                    rotten.append((x, y))
                    grid[x][y] = 2
        minutes += 1
    return max(0, minutes - 1) if not fresh else -1

File: 0994_rotting_oranges/rotting_oranges.js
// Time: O(n * m)
// Space: O(n * m)
const orangesRotting = (grid) => {
  const queue = []
  let fresh = 0
  let m = grid.length
  let n = grid[0].length
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 2) queue.push([i, j])
      else if (grid[i][j] === 1) fresh++
    }
  }
  if (fresh === 0) return 0
  let count = 0
  while (queue.length) {
    const size = queue.length
    const directions = [
      [0, 1],
      [1, 0],
      [-1, 0],
      [0, -1],
    ]
    for (let i = 0; i < size; i++) {
      const [x, y] = queue.shift()
      for (let [r, c] of directions) {
        const xc = x + r
        const yc = y + c
        if (
          xc < 0 ||
          xc >= m ||
          yc < 0 ||
          yc >= n ||
          grid[xc][yc] === 0 ||
          grid[xc][yc] === 2
        )
          continue
        grid[xc][yc] = 2
        fresh--
        queue.push([xc, yc])
      }
    }
    count++
  }
  return fresh ? -1 : count - 1
}
